src/main/java/org/jenkinsci/plugins/codesonar/services/MaxCyclomaticComplexityJsonReader.java

```diff
@@ -0,0 +1,296 @@
+package org.jenkinsci.plugins.codesonar.services;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Logger;
+
+import org.jenkinsci.plugins.codesonar.models.ProcedureMetric;
+
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+
+/**
+ * Json parser implementation specialized on processing the response of the HUB JSON API which endpoint is /analysis/{analysis}-procedures.json.
+ * In particular, its purpose is to extract the CodeSonar procedure having the maximum cyclomatic complexity.
+ * Its implementation leverages the streamed parsing capability of the Google GSon library in order to avoid loading the entire response in-memory. 
+ * @author aseno
+ *
+ */
+class MaxCyclomaticComplexityJsonReader {

```

Comment:
We had a long discussion about this class and how it is tightly coupled to `ProceduresService` and `ProcedureMetric`.  If we were to fix this, we might try to refactor this class into a generic "ProcedureRowJsonReader" which would not read an array of rows, but would be a wrapper on a JSON stream and would produce one procedure row for each call to its "read" method.  The important elements to refactor are:

1. This class should not read all rows in the JSON stream into a single in-memory array, it should instead read one row at a time, like an iterator.  Consequently, this object should not be responsible for sorting the rows.  The problem of sorting the rows should be handled by the caller.
2. This class should not be hard-wired to cyclomatic complexity metrics.

We decided that refactoring this will be time-consuming and it will not ultimately solve any immediate problems, so we are willing to postpone this work to a later release provided that we don't find any significant issues in this code.

Reply:
I found some problems with the `MaxCyclomaticComplexityJsonReader` when reading CodeSonar 7.4 JSON, so I went ahead and replaced it with a more general `ProcedureJsonRowReader`.  I moved the in-memory sorting to `ProcedureService`.  I chose the name "`ProcedureJsonRowReader`" since it is a reader of `ProcedureJsonRow` objects.

I would like to eventually move the ordering done in `ProcedureJsonRow` to a `Comparator` class, but I didn't do that.